### PR TITLE
Fix feature validation e2e on master

### PR DESCRIPTION
**Pull request by seanknox**
This PR is aimed at allowing the set of feature validations (managed disks, Calico, etc.) to run successfully on the master branch. There should be no impact on Jenkins, all changes are new files and/or for CircleCI only. 
- Break feature validations into distinct CI stages 
- Use a different subscription for each e2e category:
  - PR e2e
  - Master branch Kubernetes e2e
  - Master branch DC/OS e2e
  - Master branch Swarm e2e
  - Master branch Docker Swarm e2e
- By default CircleCI will wait 10m for output from a build before aborting. Our e2e tests sometimes hit this limit and CI then aborts the build.  Bumped`no_output_timeout` from the default 10m to 30m, as documented here: https://circleci.com/docs/2.0/configuration-reference/

---
This change is [Reviewable](https://reviewable.io/reviews/azure/acs-engine/1176)

### CODE REVIEW

**Comment by seanknox** [comment]
Update: we're now running into VM quota issues. I'm working with the quota folks to bump up our limits. Long-term we should consider building clusters that have as many feature flags enabled as possible so we can test multiple features at once, instead of building one-offs.

File: .circleci/config.yml [hunk 1]
```diff
@@ -79,7 +106,12 @@ workflows:
...
-      - master:
+      - master-kubernetes-e2e:
           filters:
             branches:
-              only: master
+              only: increase-feature-e2e-timeout
+      - master-swarm-e2e:
+          filters:
+            branches:
+              only: increase-feature-e2e-timeout
```
Comment (seanknox) [inline .circleci/config.yml]: Reminder for me to change this to `master` when this PR is good.

**Comment by seanknox** [comment]
@jackfrancis @jchauncey expanding the scope of this PR as the CI timeouts are just one of the issues preventing feature validations from running on master. As next small step, I split out the tests from `test/acse-conf/acse-feature-validation.json` by orchestrator in `test/e2e`. Circle CI allows us to chain workflows together, so now orchestrator validations can run independent of each other. Benefits of this approach:

1. Grouping each set of orchestrator functional tests makes it easier to debug when there is a failure.
1. Since a smaller set of Azure deployments are created in each phase, we are less likely to run into quota issues and have a better chance of the `acs-engine-test` runner running to completion so it can clean up after itself.
1. Allows us to further split e2e deployments across subscriptions, lessening issues with quotas.

cc @dtzar

File: .circleci/config.yml [hunk 2]
```diff
@@ -79,7 +160,28 @@ workflows:
...
+            - test
+          filters:
+            branches:
+              only: increase-feature-e2e-timeout
+      - master-swarm-e2e:
+          requires:
+            - test
+          filters:
+            branches:
+              only: increase-feature-e2e-timeout
```
Comment (seanknox) [inline .circleci/config.yml]: Reminder for me to change this back `master` when this PR is done.

File: .circleci/config.yml [hunk 3]
```diff
@@ -54,18 +55,98 @@ jobs:
...
+          echo 'export BUILD_NUM=${CIRCLE_BUILD_NUM}' >> $BASH_ENV
+          echo 'export RESOURCE_GROUP_PREFIX=y' >> $BASH_ENV
+          echo 'export SKIP_METRICS=y' >> $BASH_ENV
+          echo 'export STAGE_TIMEOUT_MIN=30' >> $BASH_ENV
+          echo 'export NUM_OF_RETRIES=2' >> $BASH_ENV
+          echo 'export SUBSCRIPTION_ID=${SUBSCRIPTION_ID_E2E_SWARM}' >> $BASH_ENV
+          echo 'export SERVICE_PRINCIPAL_CLIENT_ID=${SERVICE_PRINCIPAL_CLIENT_ID_E2E_SWARM}' >> $BASH_ENV
+          echo 'export SERVICE_PRINCIPAL_CLIENT_SECRET=${SERVICE_PRINCIPAL_CLIENT_SECRET_E2E_SWARM}' >> $BASH_ENV
+          echo 'export CLUSTER_SERVICE_PRINCIPAL_CLIENT_ID=${SERVICE_PRINCIPAL_CLIENT_ID_E2E_SWARM}' >> $BASH_ENV
+          echo 'export CLUSTER_SERVICE_PRINCIPAL_CLIENT_SECRET=${SERVICE_PRINCIPAL_CLIENT_SECRET_E2E_SWARM}' >> $BASH_ENV
```
Comment (seanknox) [inline .circleci/config.yml]: Not very DRY, but not sure of a better/clearer way to do this.

File: .circleci/config.yml [hunk 4]
```diff
@@ -54,18 +55,98 @@ jobs:
...
-          name: Feature validation tests
+          name: Kubernetes feature validation tests
+          command: make bootstrap build test-e2e
+      - store_artifacts:
+          path: /go/src/github.com/Azure/acs-engine/_logs
+  master-swarm-e2e:
+    working_directory: /go/src/github.com/Azure/acs-engine
+    no_output_timeout: "30m"
+    docker:
+      - image: microsoft/acs-engine:latest
```
Comment (jchauncey) [inline .circleci/config.yml]: What's the purpose of this image here in this part of the CI run? Does it hold the binary we are testing? If so when did it get built?

Comment (seanknox) [inline .circleci/config.yml]: > Does it hold the binary we are testing?

The image at the moment provides 1) a go dev environment and 2) the `az`CLI tool installed. Circle launches the build into a container with this image, then mounts the current source code of the branch into it, runs bootstrap, builds binaries, etc. 

I want to build immutable Docker artifacts for each build and run tests in that, but I couldn't figure out how to cache and reuse the built image between CI stages. I think it's possible, but takes some reading. Without caching a built image we'd add 3-4 minutes on every stage, so opted for the current approach in the interim.

**Comment by seanknox** [comment]
Circle CI isn't honoring the `no_output_timeout` setting for PR e2e. Going to try opening a new branch/PR to see if that clears the cache.

**Pull request closed without merging**